ood-portal-generator/lib/ood_portal_generator/application.rb

```diff
@@ -29,41 +33,214 @@ def context
         config.file? ? YAML.load_file(config) : {}
       end
 
+      def rpm
+        @rpm.nil? ? false : @rpm
+      end
+
+      def force
+        @force.nil? ? false : @force
+      end
+
+      def detailed_exitcodes
+        @detailed_exitcodes.nil? ? false : @detailed_exitcodes
+      end
+
+      def change_exit
+        detailed_exitcodes ? 3 : 0
+      end
+
+      def skip_exit
+        detailed_exitcodes ? 4 : 0
+      end
+
+      def apache
+        ENV['APACHE'] || OodPortalGenerator.scl_apache ? '/opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf' : '/etc/httpd/conf.d/ood-portal.conf'
+      end
+
+      def apache_bak
+        "#{apache}.#{Time.now.strftime('%Y%m%dT%H%M%S')}"
+      end
+
+      def apache_services
+        if OodPortalGenerator.scl_apache
+          ['httpd24-httpd', 'httpd24-htcacheclean']
+        else
+          ['httpd', 'htcacheclean']
+        end
+      end
+
+      def sum
+        ENV['SUM'] || '/etc/ood/config/ood_portal.sha256sum'
+      end
+
+      def parse_input(input)
+        lines = File.readlines(input)
+        new_lines = []
+        lines.each do |line|
+          next if line =~ /^\s*#/
+          new_lines << line
+        end
+        str = new_lines.join('')
+        str
+      end
+
+      def save_checksum(input)
+        str = parse_input(input)
+        checksum = Digest::SHA256.hexdigest(str)
+        open(sum, 'w') do |f|
+          f.puts "#{checksum} #{apache}\n"
+        end
+      end
+
+      def checksum_matches?(input)
+        checksum_str = File.readlines(sum)[0]
+        checksum = checksum_str.split(' ')[0]
+
+        str = parse_input(input)
+        new_checksum = Digest::SHA256.hexdigest(str)
+
+        checksum == new_checksum
+      end
+
+      def checksum_exists?
+        return false unless File.exist?(sum)
+        File.readlines(sum).grep(apache).size == 0
+      end
+
+      def update_replace?
+        # If checksum of ood-portal.conf matches or ood-portal.conf doesn't exit, replace is possible.
+        # If the checksum matches this means a site has not changed ood-portal.conf outside ood-portal-generator
+        if checksum_matches?(apache)
+          replace = true
+        else
+          replace = false
+        end
+        replace = true if force
+        return replace
+      end
+
+      def exit!(code)
+        case code
+        when TrueClass
+          exit(0)
+        when FalseClass
+          exit(1)
+        else
+          exit(code)
+        end
+      end
+
+      def update_ood_portal
+        ret = 0
+        changed = false
+        new_apache = Tempfile.new('new_apache')
+        @output = new_apache.path
+        self.start('generate')
```

Comment:
This calls the generate script which loads another OptParser again. Ideally we would process CLI args in one location, and help for those args would be presented to the user (so if update_ood_portal allows users to set template, output, etc. for generate command, those should be options provided). 

In that way you could imagine a separate generate method that just had a body like this:

```ruby
def generate
  view = View.new(context)
  rendered_template = view.render(template.read)
  output.write(rendered_template)
end
```

and that would be called from here and from the start command

Reply:
This has been implemented